### ISSUE TITLE
Switch to ReinforceJS

### ISSUE THREAD

**Issue by mryellow** [mryellow/ros-dqn-rat-js#1]
http://cs.stanford.edu/people/karpathy/reinforcejs/

Multi-agent prototype goal seeking.
https://github.com/mryellow/reinforcejs/tree/demo-multiagent

**Comment by sunt40** [comment]
I want ratslam's experience map node to subscribe to "corlor" topic, like the "get distance service" you added to "main_em.cpp" file. How to do ? @mryellow

**Comment by mryellow** [comment]
> "main_em.cpp"

https://github.com/mryellow/ratslam/commit/b5af3b70801f399047aa505905e2bcf573247458

hmmm `dijkstra_distance_between_experiences` already existed and I just exposed it I believe.

Defined here:

https://github.com/mryellow/ratslam/blob/304673ced121c83c7291c344d371cd9ad1482b50/src/ratslam/experience_map.cpp#L240

Over here I have an example robot definition (`indigo`) which connects the dots.

https://github.com/kulbu/kulbu_slam/blob/1af3e5b0c2b2990d1cbb139d2d8e575eaba084af/launch/rat.launch

It's been awhile but this looks to do something with the goal.

```
    <node name="RatSLAMExperienceMap" pkg="ratslam_ros" type="ratslam_em" args="$(find kulbu_slam)/config/rat.txt _image_transport:=compressed" cwd="node" required="true">
      <!-- FIXME: use `cmd_vel` -->
      <remap from="/ratslam/ExperienceMap/CmdToGoal" to="$(arg topic_cmd)"/>
    </node>
```

Also some plots which seem related. Believe this is my debugging view, not as completely integrated as what you were hoping.

```
    <node if="$(arg show_plot_distance)" pkg="rqt_plot" type="rqt_plot" name="plot_dm" args="/rtab_rat/DistanceMatrix/a,/rtab_rat/DistanceMatrix/b,/rtab_rat/DistanceMatrix/c,/rtab_rat/DistanceMatrix/d,/rtab_rat/DistanceMatrix/e,/rtab_rat/DistanceMatrix/f,/rtab_rat/DistanceMatrix/g,/rtab_rat/DistanceMatrix/h,/rtab_rat/DistanceMatrix/i"/>

    <node if="$(arg show_plot_distance)" pkg="rqt_plot" type="rqt_plot" name="plot_dm_d_e" args="/rtab_rat/DistanceMatrix/calc1,/rtab_rat/DistanceMatrix/calc2"/>
```

`rtab_rat` is a hint towards my original thinking, was aiming for an agent which could create decent RTAB-Map results, exploring areas and viewing them from multiple directions. I'm not sure exactly where the node is that's meant to be connecting to.

Have a feeling this part was written, then disabled, then the rest of the code-base diverged from it. Could be `rtab_rat` was renamed simply to `ratslam`.

**Comment by mryellow** [comment]
Yeah really struggling to find what it is I had that `show_plot_distance` connected to. `rtab` suggests something to do with RTAB-Map but think it might still be in reference to the distance calculation.

Suffice to say it's something to do with debugging the internals of some C++ code somewhere.

There was a whole bit where I was looking at the difference between a goal and a sub-goal too.

**Comment by sunt40** [comment]
emm...
I just want the experience map node to subscribe to one more object information node and then dot the detected object on the map.

**Comment by mryellow** [comment]
Yeah I didn't end up going the route of plotting my goal or subgoal onto the map or I'd have more for you to follow on that path. 

Instead I was using that simple X/Y plot for debugging to avoid having to dig into that part of things.

**Comment by sunt40** [comment]
yes，emm
Can you give me some advice on the combination of ratslam and object detection ？

**Comment by mryellow** [comment]
Not sure how it would work exactly.

That experience map is a temporal map, directions and driving times. It kind of re-jigs itself all the time so the nodes are the right distance from each other. I'm not really certain if it has a strong X/Y coordinate type system where you could say "Object is at X/Y, render it on the map".

You might need more "Object is 3 seconds in this direction from that other node" type mapping.

**Comment by mryellow** [comment]
I guess you'd just store the object like any other node in the map, as if it were a place the agent had visited. However then give it a `type` so you can then render the line a different colour or something.

**Comment by sunt40** [comment]
You are right.That's what I think. Is there a similar code to refer to?

**Comment by mryellow** [comment]
Not that I've got, it was a touch too much work at the time. Didn't dig too deeply into the whole experience map display.
